debian/pysdm.8
.TH PYSDM 8
.SH NAME
pysdm \- PyGTK Storage Device Manager
.SH SYNOPSIS
.B pysdm
.br
.SH DESCRIPTION
PySDM is a Storage Device Manager that allows full customization of hard disk mountpoints without manually access to fstab.
.SH SEE ALSO
.BR mount (8),
.BR umount (8),
.BR fstab(5).
.br
.SH AUTHOR
pysdm was written by Jaime Soriano Pastor
.PP
This manual page was written by Jaime Soriano Pastor <kronoss@kronoss.org>,
for the Ubuntu project (but may be used by others).


debian/changelog
pysdm (0.4.1-0blankon1) rote; urgency=low

  * Initial Release.

 -- mijortsa <azis.astrojim@surabaya.di.blankon.in>  Thu, 16 Aug 2012 21:02:24 +0700

pysdm (0.4.1-0ubuntu1) hardy; urgency=low

  [ Cody A.W. Somerville ]
  * New upstream release:
    - Some GUI changes for better GNOME HIG compliance
    - Show a dialog when not root user and close
    - Confirmation dialog added when removing udev rule
    - Fixed pysdm.desktop thanks to Emmet Hikory
    - Fixed is_mountable() and get_vfstype() functions in blkid.py
    - Fixed block.dev detection by major minor
  * debian/compat: Bumped to level 5
  * debian/control:
    - Bumped standards version
    - Moved debhelper build-dep-indep to build-dep
    - Bumped debhelper build-dep to version 5
    - Added build-dep-indep on python-support
    - Added homepage field
    - Added binary package depends on python-gtk2, and python-glade2
  * deban/rules:
    - Cleaned up
    - Removed -i flags from all calls (not needed)
    - Fixed make-clean-ignores-errors
    - Updated to use python-support instead of dh_python
    - Removed call to dh_installmapages
    - Updated call to dh_installman to actually work
  * Closes lp: #64832

  [ Emmet Hikory ]
  * Update the .desktop file patch to meet updated standards
  * Remove spurious shebang lines
  * Set Recommends: on gksu for .desktop file

 -- Emmet Hikory <persia@ubuntu.com>  Sat, 19 Jan 2008 18:13:30 +0900

pysdm (0.4-1ubuntu3) hardy; urgency=low

  * Added watch file (LP: #180390)

 -- Steve Stalcup <vorian@ubuntu.com>  Fri, 04 Jan 2008 12:25:29 -0500

pysdm (0.4-1ubuntu2) feisty; urgency=low

  * Rebuild for python2.5 as the default python version.

 -- Matthias Klose <doko@ubuntu.com>  Fri, 12 Jan 2007 13:20:58 +0000

pysdm (0.4-1ubuntu1) dapper; urgency=low

  * Edited the desktop file to validate

 -- Emmet Hikory <emmet.hikory@gmail.com>  Mon,  8 May 2006 11:56:07 +0900

pysdm (0.4-1) breezy; urgency=low

  * File systems can be removed from fstab
  * Partitions whithout fstab entry can have udev rules now
  * Preserve fstab comments and empty lines
  * File systems can be unmounted now
  * (no)diratime option added
  * Fixed some fstab formatting bugs
  * Fixed blkid caching problem
  * Fixed default options
  * Fixed XFS quota options 
  * Fixed exception when all default options are selected
  * Spanish l10n updated

 -- Jaime Soriano Pastor <kronoss@kronoss.org>  Sat,  3 Sep 2005 18:33:04 +0200

pysdm (0.3.9.1-1) unstable; urgency=low

  * udev management

 -- Jaime Soriano Pastor <kronoss@kronoss.org>  Fri,  2 Sep 2005 23:43:24 +0200

pysdm (0.3.9-1) unstable; urgency=low

  * udev management (unstable)

 -- Jaime Soriano Pastor <kronoss@kronoss.org>  Fri,  2 Sep 2005 17:04:48 +0200

pysdm (0.3.1-1) unstable; urgency=low

   * Fixed formatting of fstab
   * Fixed mounting of new configured partitions
   * Fixed /none bug
   * Mount button is named "Remount" if file system is mounted
   * Actualized spanish l10n

 -- Jaime Soriano Pastor <kronoss@kronoss.org>  Thu,  1 Sep 2005 12:21:19 +0200

pysdm (0.3-1) unstable; urgency=low

  * i18n and l10n support
  * spanish l10n
  * mount button
  * root filesystem options configurable

 -- Jaime Soriano Pastor <kronoss@kronoss.org>  Wed, 31 Aug 2005 15:10:03 +0200

pysdm (0.2.1-1) unstable; urgency=low

  * User/Advanced tabbed options dialog
  * Fixed default button bug
  * Support for ext2, ext3, fat, vfat, msdos, ntfs, reiserfs, swap and xfs 
    special options

 -- Jaime Soriano Pastor <kronoss@kronoss.org>  Tue, 30 Aug 2005 01:19:21 +0200

pysdm (0.2-1) unstable; urgency=low

  * PySDM doesn't destroy now fstab
  * New options interface
  * Changed install dir to /usr/share/pysdm

 -- Jaime Soriano Pastor <kronoss@kronoss@kronoss.org>  Mon, 29 Aug 2005 14:43:48 +0200

pysdm (0.1-1) unstable; urgency=low

  * Initial release.

 -- Jaime Soriano Pastor <kronoss@kronoss.org>  Wed, 24 Aug 2005 14:44:40 +0200



pysdm/pysdm.py
# -*- coding: UTF8 -*-

# Python module pysdm.py
# Autogenerated from pysdm.glade
# Generated on Thu May 11 20:01:12 2006

# Warning: Do not modify any context comment such as #--
# They are required to keep user's code

import os
import gtk
import re
import thread

import constants
import posix
from options_ui import *
from blkid import *
from fstab import *
from sysfs import *

from SimpleGladeApp import SimpleGladeApp
from SimpleGladeApp import bindtextdomain

app_name = "pysdm"
app_version = "0.3"

glade_dir = ""
locale_dir = ""

import gettext
gettext.install(app_name)
#bindtextdomain(app_name, locale_dir)

PARTITIONS = "/proc/partitions"
FSTAB = "/etc/fstab"
DEFAULT_MOUNT = "/media"
UDEV_DIR = "/etc/udev/"
UDEV_RULES = UDEV_DIR + "user.rules"

class Mainwindow(SimpleGladeApp):

    def __init__(self, path = os.path.join (constants.data_dir, "pysdm.glade"),
                 root="mainWindow",
                 domain=app_name, **kwargs):
        path = os.path.join(glade_dir, path)
        SimpleGladeApp.__init__(self, path, root, domain, **kwargs)

    #-- Mainwindow.new {
    def new(self):
        self.main_widget.set_title(_("Storage Device Manager"))
        #Load fstab data
        self.fstab = Fstab(FSTAB)
        self.urules = UdevFile(UDEV_RULES)
        self.rules = []
        self.blocks = {}

        self.current_FS = None

        #Initializes partition tree
        ptree = self.get_widget("partitiontree")
        column = gtk.TreeViewColumn(_("Partition List"), gtk.CellRendererText(), text=0)
        ptree.append_column(column)
        self.refresh_partitions(self, None)

        #Initializes rules view
        rulesv = self.get_widget("rules_treeview")
        column = gtk.TreeViewColumn(_("Rules"), gtk.CellRendererText(), text=0)
        rulesv.append_column(column)

        self.get_widget("frame3").set_sensitive(False)
    #-- Mainwindow.new }

    #-- Mainwindow custom methods {
    def refresh_panel(self, partition):
        """
        Update General Information for the partition given

        partition:
            The name of the partition in /dev tree
        """
        filesystem = self.fstab.getFS(partition, None)
        name = re.split('/', filesystem.file)
        name = name[len(name)-1]

        name_entry = self.get_widget("name_entry")
        mountpoint_entry = self.get_widget("mountpoint_entry")
        type_entry = self.get_widget("type_entry")

        self.get_widget("main_notebook").set_sensitive(True)
        self.get_widget("frame3").set_sensitive(True)

        #Root filesystem cannot be deleted
        if filesystem.file == "/":
            self.get_widget("remove_fs_button").set_sensitive(False)
        else:
            self.get_widget("remove_fs_button").set_sensitive(True)

        if filesystem.file == "/" or cmp(filesystem.file, "none")==0:
            self.get_widget("name_entry").set_sensitive(False)
            self.get_widget("openmountpoint").set_sensitive(False)
            self.get_widget("mount_button").set_sensitive(False)
        else:
            self.get_widget("name_entry").set_sensitive(True)
            self.get_widget("openmountpoint").set_sensitive(True)
            self.get_widget("mount_button").set_sensitive(True)
        
        mount_label = self.get_widget("mount_button").get_child().get_child().get_children()[1]
        if(filesystem.is_mounted()):
            mount_label.set_text(_("Unmount"))
        else:
            mount_label.set_text(_("Mount"))

        name_entry.set_text(name)
        mountpoint_entry.set_text(filesystem.file)
        type_entry.set_text(filesystem.vfstype)

        self.refresh_udev_panel(partition)

    def refresh_udev_panel(self, partition):
        block = self.blocks[partition]
        for att_name in block.attributes:
            if block.attributes[att_name]!=None:
                self.get_widget("sys_" + att_name + "_label").set_text(block.attributes[att_name])
            else:
                self.get_widget("sys_" + att_name + "_label").set_text(_("Unknown"))      
    
    def get_options(self):
        options_str = self.get_widget("options_entry").get_text()
        if len(options_str)==0: options_str = "defaults"
        return re.compile("[,\s]*").split(options_str)

    def set_options(self, options_array):
        if len(options_array)==0: options_array=["defaults"]
        options_str = options_array[0]
        for option in options_array[1:]:
            options_str = options_str + "," + option
        self.get_widget("options_entry").set_text(options_str)
        self.current_FS.mntops = options_array

    def refresh_rules(self, rules_array):
        rtree = self.get_widget("rules_treeview")

        model = gtk.TreeStore(str)
        self.rules = []

        for rule in rules_array:
            row = rule.__repr__()
            row = (row[:len(row)-1],)
            iter = model.append(None, row)
            self.rules.append(rule)
        rtree.set_model(model)

    def auto_configure(self, partition):
        """
        Autoconfigures a new partition

        partition:
            The name of the partition in /dev tree to autoconfigure

        """
        vfstype = get_vfstype(partition)
        name = re.split('/', partition)
        name = name[len(name)-1]
        self.fstab.addFS(Filesystem(partition, DEFAULT_MOUNT + "/" + name, vfstype, ["defaults"], "0", "0"))


    def get_partitions(self):
        """
        Find the partitions in the system
        """

        self.blocks = {}
        partitions = []
        partitions_file = open(PARTITIONS, "r")
        lines = partitions_file.readlines()
        for line in lines[2:]:
            partition = re.split('\s+', line)[4]
            device = re.compile("([a-z]+)([0-9]+)?").match(partition).groups()[0]

            if partition == device:
                block_path = "/sys/block/" + device
            else:
                block_path = "/sys/block/" + device + "/" + partition

            try:
                os.stat(block_path)
                #os.stat("/dev/" + partition)
                partitions.append(partition)
                block = Block(partition, device)

                self.blocks[block.name] = block
                self.blocks[block.dev] = block
            except OSError:
                pass
            except AttributeError:
                print "Not suitable dev for " + block.name

        return partitions
    #-- Mainwindow custom methods }

    #-- Mainwindow.refresh_partitions {
    def refresh_partitions(self, widget, *args):
        """
        Updates partition tree
        """

        ptree = self.get_widget("partitiontree")
        partitions = self.get_partitions()
        model = gtk.TreeStore(str)
        lastdevice = partitions[0]
        iter = model.append(None, (partitions[0],))

        for partition in partitions[1:]:
            device = partition[:len(partition)-1]
            if lastdevice != device:
                lastdevice = partition
                iter = model.append(None, (partition,))
            else:
                if is_mountable(self.blocks[partition].dev):
                    model.append(iter, (partition,))

        ptree.set_model(model) 
    #-- Mainwindow.refresh_partitions }

    #-- Mainwindow.on_partitiontree_cursor_changed {
    def on_partitiontree_cursor_changed(self, widget, *args):
        ptree = self.get_widget("partitiontree")
        selection = ptree.get_selection()
        model, selected = ptree.get_selection().get_selected()

        #User has selected a device
        if len(ptree.get_selection().get_selected_rows()[1][0])==1: 
            return

        partition = self.blocks[model.get_value(selected, 0)].dev

        #If exists a configuration in FSTAB
        if self.fstab.getFS(partition, None) != None:
            self.current_FS = self.fstab.getFS(partition, None)
            self.refresh_panel(partition)
            self.set_options(self.current_FS.mntops)
        else:
            ask_configure = gtk.Dialog(_("Configure now?"), self.main_widget, gtk.DIALOG_MODAL | gtk.DIALOG_DESTROY_WITH_PARENT, (gtk.STOCK_OK, gtk.RESPONSE_ACCEPT, gtk.STOCK_CANCEL, gtk.RESPONSE_REJECT))
            ask_configure.vbox.pack_start(gtk.Label("\n" + partition + _(" hasn't been configured.\nDo you want to configure it now?\n")))
            ask_configure.show_all()
            result = ask_configure.run()
            ask_configure.destroy()
            if result == gtk.RESPONSE_ACCEPT:
                self.auto_configure(partition)
                self.current_FS = self.fstab.getFS(partition, None)
                self.refresh_panel(partition)
                self.set_options(self.current_FS.mntops)
            else:
                self.current_FS = Filesystem(partition, "", "", "", "", "")
                self.get_widget("main_notebook").set_sensitive(True)
                self.get_widget("frame3").set_sensitive(False)
                self.refresh_udev_panel(partition)

        self.get_widget("status_bar").push(0, self.current_FS.spec)
        self.refresh_rules(self.urules.get_matching_rules(self.blocks[partition]))

    #-- Mainwindow.on_partitiontree_cursor_changed }

    #-- Mainwindow.on_name_entry_changed {
    def on_name_entry_changed(self, widget, *args):
        if cmp(self.current_FS.file, "none")==0:
            return
        path = re.split("/", self.current_FS.file)
        npath = ""
        for folder in path[1:len(path)-1]:
            npath = npath + "/" + folder
        
        npath = npath + "/" +  widget.get_text()
        self.get_widget("mountpoint_entry").set_text(npath)
        
        self.current_FS.file = npath
    #-- Mainwindow.on_name_entry_changed }

    #-- Mainwindow.on_openmountpoint_clicked {
    def on_openmountpoint_clicked(self, widget, *args):
        filechooser = gtk.FileChooserDialog(_("Select a file for mountpoint..."), self.main_widget, action=gtk.FILE_CHOOSER_ACTION_SELECT_FOLDER, buttons=(gtk.STOCK_OPEN, gtk.RESPONSE_ACCEPT, gtk.STOCK_CANCEL, gtk.RESPONSE_REJECT))
        filechooser.set_modal(True)
        filechooser.set_current_folder(DEFAULT_MOUNT)
        filechooser.show()
        response = filechooser.run()
        filechooser.hide()
        if response == gtk.RESPONSE_ACCEPT:
            self.get_widget("mountpoint_entry").set_text(filechooser.get_filename())
            self.current_FS.file = filechooser.get_filename()
        
            name = re.split('/', self.current_FS.file)
            name = name[len(name)-1]
            self.get_widget("name_entry").set_text(name)

        filechooser.destroy()
    #-- Mainwindow.on_openmountpoint_clicked }

    #-- Mainwindow.on_options_entry_changed {
    def on_options_entry_changed(self, widget, *args):
        self.current_FS.mntops = self.get_options()
    #-- Mainwindow.on_options_entry_changed }

    #-- Mainwindow.on_defaults_button_clicked {
    def on_defaults_button_clicked(self, widget, *args):
        self.set_options(["defaults"])
    #-- Mainwindow.on_defaults_button_clicked }

    #-- Mainwindow.on_open_options_clicked {
    def on_open_options_clicked(self, widget, *args):
        dialog = option_dialog(self.current_FS.vfstype, title=_("Select options"), flags=gtk.DIALOG_MODAL | gtk.DIALOG_DESTROY_WITH_PARENT, buttons=(gtk.STOCK_OK, gtk.RESPONSE_ACCEPT, gtk.STOCK_CANCEL, gtk.RESPONSE_REJECT))
        dialog.set_value(self.current_FS.mntops)
        if dialog.run() == gtk.RESPONSE_ACCEPT:
            self.set_options(dialog.get_value())
        dialog.destroy()
    #-- Mainwindow.on_open_options_clicked }

    #-- Mainwindow.on_mount_button_clicked {
    def on_mount_button_clicked(self, widget, *args):
        try:
            os.makedirs(self.current_FS.file)
        except OSError:
            pass

        if self.current_FS.is_mounted():
            mount_ret = self.current_FS.umount()

            if mount_ret!=0:
                if mount_ret==256:
                    error_message = _("The file system is busy.\nCan't be unmounted")
                else:
                    error_message = _("The file system can't be mounted\nmount returned error:") + str(mount_ret)
                dialog = gtk.Dialog(_("Mount error"), self.main_widget, gtk.DIALOG_MODAL | gtk.DIALOG_DESTROY_WITH_PARENT, (gtk.STOCK_OK, gtk.RESPONSE_ACCEPT))
                dialog.vbox.set_border_width(10)
                dialog.vbox.pack_start(gtk.Label(error_message))
                dialog.show_all()
                dialog.run()
                dialog.destroy()
        else:
            mount_ret = self.current_FS.mount()

        mount_label = self.get_widget("mount_button").get_child().get_child().get_children()[1]
        if(self.current_FS.is_mounted()):
            mount_label.set_text(_("Unmount"))
        else:
            mount_label.set_text(_("Mount"))
            

    #-- Mainwindow.on_mount_button_clicked }

    #-- Mainwindow.on_remove_fs_button_clicked {
    def on_remove_fs_button_clicked(self, widget, *args):
        dialog = gtk.MessageDialog(self.main_widget, gtk.DIALOG_MODAL | gtk.DIALOG_DESTROY_WITH_PARENT, gtk.MESSAGE_WARNING, gtk.BUTTONS_OK_CANCEL, _("This action will delete file system configuration\nAre you sure?"))
        dialog.show_all()
        if dialog.run()==gtk.RESPONSE_OK:
            self.fstab.delFS(self.current_FS.spec, self.current_FS.file)
            self.get_widget("frame3").set_sensitive(False)
        dialog.destroy()
    #-- Mainwindow.on_remove_fs_button_clicked }

    #-- Mainwindow.on_rules_treeview_cursor_changed {
    def on_rules_treeview_cursor_changed(self, widget, *args):
        pass
    #-- Mainwindow.on_rules_treeview_cursor_changed }

    #-- Mainwindow.on_udev_up_button_clicked {
    def on_udev_up_button_clicked(self, widget, *args):
        try:
            selected = self.get_widget("rules_treeview").get_selection().get_selected_rows()[1][0][0]
        except IndexError:
            return
        rule1 = selected
        rule2 = selected - 1
        if rule2 < 0: return
        self.urules.switch_rule(self.rules[rule1].line, self.rules[rule2].line)
        self.refresh_rules(self.urules.get_matching_rules(self.blocks[self.current_FS.spec]))
    #-- Mainwindow.on_udev_up_button_clicked }

    #-- Mainwindow.on_udev_down_button_clicked {
    def on_udev_down_button_clicked(self, widget, *args):
        try:
            selected = self.get_widget("rules_treeview").get_selection().get_selected_rows()[1][0][0]
        except IndexError:
            return
        rule1 = selected
        rule2 = selected + 1
        if rule2 > len(self.rules): return
        self.urules.switch_rule(self.rules[rule1].line, self.rules[rule2].line)
        self.refresh_rules(self.urules.get_matching_rules(self.blocks[self.current_FS.spec]))

    #-- Mainwindow.on_udev_down_button_clicked }

    #-- Mainwindow.on_udev_edit_button_clicked {
    def on_udev_edit_button_clicked(self, widget, *args):
        try:
            selected = self.get_widget("rules_treeview").get_selection().get_selected_rows()[1][0][0]
        except IndexError:
            return
        rule_line = self.rules[selected].line
        rules = self.urules.rules[rule_line]
        block = self.blocks[self.current_FS.spec]

        dialog = RulesMakerDialog(block, rules, title=_("New rule..."), flags=gtk.DIALOG_MODAL | gtk.DIALOG_DESTROY_WITH_PARENT, buttons=(gtk.STOCK_OK, gtk.RESPONSE_ACCEPT, gtk.STOCK_CANCEL, gtk.RESPONSE_REJECT))
        dialog.show_all()
        if dialog.run() == gtk.RESPONSE_ACCEPT:
            self.urules.set_rule(rule_line, dialog.get_rule())
        dialog.destroy()

        self.refresh_rules(self.urules.get_matching_rules(block))
    #-- Mainwindow.on_udev_edit_button_clicked }

    #-- Mainwindow.on_udev_delete_button_clicked {
    def on_udev_delete_button_clicked(self, widget, *args):
        try:
            selected = self.get_widget("rules_treeview").get_selection().get_selected_rows()[1][0][0]
        except IndexError:
            return
        dialog = gtk.MessageDialog(self.main_widget, gtk.DIALOG_MODAL | gtk.DIALOG_DESTROY_WITH_PARENT, gtk.MESSAGE_WARNING, gtk.BUTTONS_OK_CANCEL, _("Remove selected udev rule?"))
        dialog.show_all()
        if dialog.run()==gtk.RESPONSE_OK:
            rule_line = self.rules[selected].line
            self.urules.del_rule(rule_line)
            self.refresh_rules(self.urules.get_matching_rules(self.blocks[self.current_FS.spec]))
        dialog.destroy()
    #-- Mainwindow.on_udev_delete_button_clicked }

    #-- Mainwindow.on_udev_new_button_clicked {
    def on_udev_new_button_clicked(self, widget, *args):
        dialog = RulesMakerDialog(self.blocks[self.current_FS.spec], title=_("New rule..."), flags=gtk.DIALOG_MODAL | gtk.DIALOG_DESTROY_WITH_PARENT, buttons=(gtk.STOCK_OK, gtk.RESPONSE_ACCEPT, gtk.STOCK_CANCEL, gtk.RESPONSE_REJECT))
        dialog.show_all()
        if dialog.run() == gtk.RESPONSE_ACCEPT:
            self.urules.add_rule(dialog.get_rule())
        dialog.destroy()
        self.refresh_rules(self.urules.get_matching_rules(self.blocks[self.current_FS.spec]))
    #-- Mainwindow.on_udev_new_button_clicked }

    #-- Mainwindow.on_apply_clicked {
    def on_apply_clicked(self, widget, *args):
        os.rename(FSTAB, FSTAB + ".BAK")
        for filesystem in self.fstab.filesystems:
            if type(filesystem) == str: continue
            if cmp(filesystem.file,"none")!=0:
                try:
                    os.makedirs(filesystem.file)
                except OSError:
                    pass
        self.fstab.toFile(FSTAB)
        self.urules.write()
    #-- Mainwindow.on_apply_clicked }


#-- main {

def main():
	if posix.getuid() != 0:
		dialog = gtk.MessageDialog(None, 0, gtk.MESSAGE_WARNING, gtk.BUTTONS_OK, _("You should have root privileges to run this program"))
		dialog.show_all()
		dialog.run()
	else:
		main_window = Mainwindow()
		main_window.run()

if __name__ == "__main__":
    main()

#-- main }
